inference-engine/src/transformations/src/transformations/serialize.cpp

```diff
@@ -93,18 +93,19 @@ class ConstantWriter {
             m_binary_output.write(ptr, size);
             return offset;
         }
-        // The biggest supported models have at maximum 1-2 thousand constant nodes,
-        // with 64 bit hash that gives a probability around 1 in 10 trillion that a
-        // hash collision will appear. Because of this, a choice has been made to
-        // not perform collision detection and keep the hashing quick and seamless.
+        // This hash is weak (but efficient) and must be replace with some other
+        // more stable hash algorithm. For example current hash algorithms gives
+        // the same hash for {2, 2} and {0, 128} arrays. So we have to compare
+        // values when finding a match in hash map.
         const HashValue hash = hash_combine(ptr, size);
         const auto found = m_hash_to_file_positions.find(hash);
-        if (found != end(m_hash_to_file_positions)) {
-            return found->second;
+        if (found != end(m_hash_to_file_positions) &&
+            memcmp(static_cast<void const*>(ptr), found->second.second, size) == 0) {
+            return found->second.first;
```

Comment:
@nosovmik can we try this and check performance impact on model caching?

Reply:
I assume that this check won't take much time as we usually compress only small arrays with attributes but not weights. Also `memcmp` is superfast :)

Reply:
Discussed model caching impact with @nosovmik in offline. So I did some measurements for `NetworkCompilationContext::computeHash` function on the model where weights compressions really works (other models are not affected by this change) and here the results:
```
Both measurement were made on the same machine with 1000 of iterations.
AVG TIME (with fix): 120.418ms
AVG TIME (no fix):   118.694ms
```
So we think that the difference in 2ms is applicable.

Reply:
Numbers looks ok. Considering that 'computeHash' takes much less time than 'ReadNetwork' (for CNNNetwork) and ImportNetwork (on plugin side) - this 2ms impact on pretty large model is acceptable